### PR TITLE
Support for python 3 (+python 2)

### PR DESCRIPTION
**Pull request by piertoni**
After applying futurize stage1 I need some help on the following:
- Update Travis CI script for Python3
- on lib/taurus/qt/qtgui/display/qled.py i changed the following line in order to pass the test:

```
#def __init__(self, parent=None, ledsize=LedSize.SMALL, ledcolor=LedColor.GREEN):
def __init__(self, parent=None, ledsize=24, ledcolor=LedColor.GREEN):
```
It's an enumeration problem as the enumeration is not starting from 0.
For what I have seen can be something related to Taurus Enumerations that must start from 0.
Surely will be straightforward for some of you that knows the codebase well.

### CODE REVIEW

**Comment by cpascual** [comment]
Hi @piertoni , thanks for the PR. 
Right now the @taurus-org/integrators  are in a rush for finishing the Jan18 release of Taurus and Sardana. Please give us a couple of weeks so that we can review this PR with the clam it deserves.

For the moment I'll mark it with "Help Wanted" in the hope of attracting attention from somebody else who may be interested in reviewing this in the meanwhile.

**Comment by cpascual** [comment]
Hi @piertoni another user (@asenchenko) is also doing some effort on supporting python3 (see #765 ).
I wonder if you are still interested in pushing for this feature, and if so, if you would agree to centralise the contributions to this in your branch. This would be more agile since all the work could go on autonomously (no need for official @taurus-org/integrators attention until the moment of actually doing a PR 
to the official develop branc with the final implementation.

Alternatively, if you all feel  more comfortable this way, we can also create a py3 branch in the official repo and all python3-related PRs could go there (but then the PRs will be delayed by the official integrators' time availability).

**Comment by cpascual** [comment]
Hi, I updated this branch with the latest changes from develop (so that it does not have merge conflicts) and I **also configured travis to do tests for python3**

**Comment by cpascual** [comment]
I merged the changes from #765 here. and reverted the workaround on qled.py mentioned in the first post of the commit

**Comment by cpascual** [comment]
The testsuite is erroring now for the python3 case: see: https://travis-ci.org/taurus-org/taurus/jobs/397882052

**Comment by cpascual** [comment]
> Ok for using my branch and for the help of @asenchenko or any other.

Great, so let's do that.

Therefore, @asenchenko, please propose your PRs directly to this branch (`piertoni:futurize1`)

File: lib/taurus/core/util/enumeration.py [hunk 1]
```diff
@@ -95,9 +97,9 @@ def __init__(self, name, enumList, flaggable=False, no_doc=False):
                     raise EnumException(
                         "flagable enum does not accept tuple items")
                 x, i = x
-                if not isinstance(x, (str, unicode)):
+                if not isinstance(x, (str, str)):
```
Comment (cpascual) [inline lib/taurus/core/util/enumeration.py]: All these checks of `isinstance(x, (str,str))` and `isinstance(ix, (nt,int))` can now be simplified

File: lib/taurus/qt/qtcore/configuration/configuration.py [hunk 2]
```diff
@@ -403,7 +408,7 @@ def createQConfig(self):
         .. seealso:: :meth:`restoreQConfig`
         '''
         from taurus.external.qt import Qt
-        import cPickle as pickle
+        import pickle as pickle
```
Comment (cpascual) [inline lib/taurus/qt/qtcore/configuration/configuration.py]: simplify `import pickle as pickle`

File: lib/taurus/qt/qtgui/graphic/taurusgraphic.py [hunk 3]
```diff
@@ -26,9 +26,17 @@
...
 from __future__ import print_function
+from __future__ import division
 
 # TODO: Tango-centric
 
+from future import standard_library
+standard_library.install_aliases()
+from builtins import str
+from builtins import range
+from past.builtins import basestring
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/graphic/taurusgraphic.py]: we should later on check all the `from past.builtins` imports and see if we can write more natural code

File: lib/taurus/core/util/enumeration.py (hunk 1 above)
Comment (asenchenko) [reply lib/taurus/core/util/enumeration.py]: I suggest to use future.utils.string_type and future.utils.integer_types for type checking.
It allows us to keep code compatible with python2 and python3 .

**Comment by piertoni** [comment]
I don't understand which is the problem that causes a lot of test to fail on python3 docker.
Seems like the connection to the test devices tangoschemetest/unittest/temp-1 does not work.
Running the docker machine with the port 10000 open I can connect with Jive and find the DeviceServer sys/tg_test/1 that is working, but I can't connect to tangoscheme/unittest/temp-1
Any Idea?

```
File "/usr/lib/python3/dist-packages/tango/device_proxy.py", line 330, in __DeviceProxy__read_attribute
    return __check_read_attribute(self._read_attribute(value, extract_as))


PyTango.DevFailed: DevFailed[
DevError[
    desc = Failed to connect to device tangoschemetest/unittest/temp-1
           The connection request was delayed.
           The last connection request was done less than 1000 ms ago
  origin = Connection::reconnect
  reason = API_CantConnectToDevice
severity = ERR]
DevError[
    desc = Failed to read_attribute on device tangoschemetest/unittest/temp-1, attribute string_scalar_ro
  origin = DeviceProxy::read_attribute()
  reason = API_AttributeFailed
severity = ERR]
]
```

**Comment by asenchenko** [comment]
I was able to pass almost all test on python3 (python3.5, python 3.6) except taurus.qt.qtgui.extra_* and taurus.qt.qtgui.plot. These modules depend on Qwt.  Should  we add extra code to exclude these tests on python3?

I think we are close enough to run some application on python 3. 
Could we create some kind of road-map for python3?

**Comment by piertoni** [comment]
I think that adding code to pass Qwt tests is a good idea, let's wait for @cpascual .
The #452 will implement plots with pyqtgraph and maybe we could merge also this PR on this branch.

**Comment by cpascual** [comment]
Sorry. I am currently out of office. I'll be back on the 16th of Aug

**Comment by cpascual** [comment]
Hi, I am doing tests with this branch. I am impressed how much it advanced. Thanks for the work!

>  These modules depend on Qwt. Should we add extra code to exclude these tests on python3?

I agree with skipping the tests that depend on Qwt5 (since Qwt5 is not available for py3). I'll try to implement it and push ASAP

> The #452 will implement plots with pyqtgraph and maybe we could merge also this PR on this branch.

I think that is not needed. The pyqtgraph-based plots implementation was moved to [an independent module](https://github.com/taurus-org/taurus_pyqtgraph) that if installed is automatically used as a plugin by taurus.

So what we need in order to have pyqtgraph plots  in python3 is to futurize the taurus_pyqtgraph module itself (I expect it to be a lot easier) than for taurus

Apart from that, we should merge the latest changes from the taurus develop branch to keep this one up-to date. I'll have a look at it and I'll try to solve the merge conflicts in this week.

**Comment by cpascual** [comment]
I did the merge. Right now only a few tests are failing for py3 (mostly related with unavailable dependencies, pyqwt among them):

[Ran 636 tests in 168.241s   FAILED (failures=1, errors=5, skipped=5)](https://travis-ci.org/taurus-org/taurus/jobs/418693773)

I'll try to handle those.

**Comment by cpascual** [comment]
Steps 2 and 3 of the roadmap are already achieved after  db9b528 
Step 1 still giving trouble, though

**Comment by cpascual** [comment]
Step 1 is now fixed with c7f484b ...

So all CI tests are green now for this branch and we can go for the code review (step 4 in the [roadmap](https://github.com/taurus-org/taurus/pull/703#issuecomment-415350236) ).

I'll get to it in a couple of weeks , but if @piertoni  or @asenchenko  or anyone else wants to advance some work, you are welcome)

**Comment by piertoni** [comment]
Let's do the job!

**Comment by piertoni** [comment]
I am using my repository to track issues of the futurize branch.
What do you think to share this approach?

**Comment by asenchenko** [comment]
I think we should keep PY3 related issue in different place until step 6. It will help to avoid confusion.

File: lib/taurus/qt/qtgui/display/tauruslabel.py [hunk 4]
```diff
@@ -118,8 +119,8 @@ def _updateToolTip(self, label):
         toolTip = label.getFormatedToolTip()
         if self._trimmedText:
             toolTip = u"<p><b>Value:</b> %s</p><hr>%s" %\
-                      (str(self._text, errors='replace'),
-                       str(str(toolTip), errors='replace'))
+                      (to_str(self._text, errors='replace'),
+                       to_str(toolTip, errors='replace'))
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/display/tauruslabel.py]: Why is this necessary? 
AFAIK, self._text and toolTip should never be of type bytes, right?

**Comment by piertoni** [comment]
I was testing taurusdemo that works on python2 but not on python3 docker machine.
To replicate the error just launch taurusdemo from inside the python3-stretch docker machine and select the "Frame"

`Traceback (most recent call last):
  File "/home/lib/taurus/qt/qtgui/base/taurusbase.py", line 363, in filterEvent
    self.handleEvent(*evt)
  File "/home/lib/taurus/qt/qtgui/display/tauruslabel.py", line 318, in handleEvent
    ctrl.handleEvent(evt_src, evt_type, evt_value)
  File "/home/lib/taurus/qt/qtgui/base/tauruscontroller.py", line 120, in handleEvent
    self.update()
  File "/home/lib/taurus/qt/qtgui/base/tauruscontroller.py", line 139, in update
    self._updateToolTip(widget)
  File "/home/lib/taurus/qt/qtgui/display/tauruslabel.py", line 122, in _updateToolTip
    (str(self._text, errors='replace'),
TypeError: decoding str is not supported
`
I am working on a more clear and elegant solution for this conversion issue that is not the only one around.
I was thinking to refactor the function to_str to another named to_unicode (wich is more understandable) to be a helper function that works on both py2 and 3.

With this modification the tooltip on the tauruslabel works in both py2 and 3.
What do you think?

**Comment by cpascual** [comment]
Just some updates... 
- I am in the middle of the full code review (step 4), including removal of non-idiomatic constructions
- regarding the [py2and3 module and the to_str function](https://github.com/taurus-org/taurus/pull/703#issuecomment-420611677), I would prefer not to use it. Instead we should make sure that we are never getting bytes in those parts of the code (which AFAIK is the case) and get rid of the decoding part. I will do it in my commits for step 4.

**Comment by cpascual** [comment]
Hi, I finished with the step 4 of the [roadmap](https://github.com/taurus-org/taurus/pull/703#issuecomment-415350236).

It would be nice now that as much people as possible would test the resulting code both with py2 and with py3 (in steps 5 and 7 we will use the [checklist](https://github.com/taurus-org/taurus/blob/develop/doc/how_to_release.md#manual-test-checklist) that we use for releases, but still, there are many things that may not appear in that checklist).

**Comment by cpascual** [comment]
Hi  to @piertoni , @asenchenko , @taurus-org/integrators and everybody who is interested, this is a call for help in testing this PR **under python2** ( [step 5 of the roadmap](https://github.com/taurus-org/taurus/pull/703#issuecomment-415350236) ). 

We at alba will test under

**Comment by piertoni** [comment]
I will do testing on windows and also on linux (debian 8 for the moment), I will report Issues on my repositor if you agree.

**Comment by cpascual** [comment]
> I will do testing on windows and also on linux (debian 8 for the moment), 

Thanks @piertoni !

> I will report Issues on my repositor if you agree.

I propose to open a separate issue for reporting each set of tests (similar to what we did in [the last release](https://github.com/taurus-org/taurus/pull/784)). If you want to use your repo for the issues, that's fine for me, but using the official repo is also ok.

**Comment by asenchenko** [comment]
Hi, I'm currently busy. I could test on Ubuntu 18.04 LTS and debian 9.

**Comment by cpascual** [comment]
The testing effort has started (and some bugs alreay fixed) Here are the issues with the test reports:

- debian9 py2 ALBA #816 
- debian9 py2 DESY: #817 

The issues being found *and which are exclusive to this PR* are reported in @piertoni 's tracker: https://github.com/piertoni/taurus/issues

BTW: @piertoni , have you done any tests already on win?

**Comment by jkotan** [comment]
have you try to tests Encoded tango attributes with taurus. It looks like currently they does not work in python3-tango, i.e.: tango-controls/pytango#216 and tango-controls/pytango#213

**Comment by cpascual** [comment]
Hi @jkotan , thanks for the tip. I wasn't aware of those bugs. AFAIK, we haven't tested yet (there is no such test in our automated tests nor on our [manual checklist](https://github.com/taurus-org/taurus/blob/develop/doc/how_to_release.md#manual-test-checklist)).

We will do ASAP, but if you can try it yourself and report here, It would be of great help.

Note that at this moment we are concentrating ensuring no regressions when using python2 (step 5 of our [roadmap](https://github.com/taurus-org/taurus/pull/703#issuecomment-415350236)). If PyTango itself is not yet handling this on python3, I guess we will be affected ourselves, but at least we should ensure that for now, we are not introducing additional problems when running on python2.

**Comment by piertoni** [comment]
Sorry I am a bit full in this period, I will do tests on Windows as soon as possible
